pkg/gardenlet/controller/managedseed/managedseed_actuator.go

```diff
@@ -0,0 +1,835 @@
+	managedSeedLogger.Infof("Deregistering shoot %s as seed", kutil.ObjectName(shoot))
+	if err := a.deregisterAsSeed(ctx, managedSeed, shoot, managedSeed.Spec.Seed != nil); err != nil {
+		return fmt.Errorf("could not dereigster shoot %q as seed: %+v", kutil.ObjectName(shoot), err)
+	}
+
+	if managedSeed.Spec.Gardenlet != nil {
+		// Delete gardenlet from the shoot
+		managedSeedLogger.Infof("Deleting gardenlet from shoot %s", kutil.ObjectName(shoot))
+		if err := a.deleteGardenlet(ctx, managedSeed, shoot); err != nil {
+			return fmt.Errorf("could not delete gardenlet from shoot %q: %+v", kutil.ObjectName(shoot), err)
+		}
+	}
+
+	// Delete garden namespace from the shoot
+	managedSeedLogger.Infof("Deleting garden namespace from shoot %s", kutil.ObjectName(shoot))
+	gardenNamespace := &corev1.Namespace{ObjectMeta: metav1.ObjectMeta{Name: v1beta1constants.GardenNamespace}}
+	if err := a.shootClient.Client().Delete(ctx, gardenNamespace); client.IgnoreNotFound(err) != nil {
+		return fmt.Errorf("could not delete garden namespace from shoot %q: %+v", kutil.ObjectName(shoot), err)
+	}
+
+	return nil
+}
+
+func (a *actuator) registerAsSeed(ctx context.Context, managedSeed *seedmanagementv1alpha1.ManagedSeed, shoot *gardencorev1beta1.Shoot) error {
+	// Check if the shoot can be registered as seed
+	if shoot.Spec.DNS == nil || shoot.Spec.DNS.Domain == nil {
+		return errors.New("cannot register shoot as seed as it does not specify a domain")
+	}
+
+	// Prepare seed spec
+	seedSpec, err := a.prepareSeedSpec(ctx, managedSeed, shoot, &managedSeed.Spec.Seed.Spec)
+	if err != nil {
+		return err
+	}
+
+	// Create or update seed object
+	seed := &gardencorev1beta1.Seed{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: managedSeed.Name,
+		},
+	}
+	_, err = controllerutil.CreateOrUpdate(ctx, a.gardenClient.Client(), seed, func() error {
+		seed.OwnerReferences = []metav1.OwnerReference{
+			*metav1.NewControllerRef(managedSeed, seedmanagementv1alpha1.SchemeGroupVersion.WithKind("ManagedSeed")),
+		}
+		seed.Labels = utils.MergeStringMaps(managedSeed.Spec.Seed.Labels, map[string]string{
+			v1beta1constants.GardenRole: v1beta1constants.GardenRoleSeed,
+		})
+		seed.Annotations = managedSeed.Spec.Seed.Annotations
+		seed.Spec = *seedSpec
+		return nil
+	})
+	return err
+}
+
+func (a *actuator) deregisterAsSeed(ctx context.Context, managedSeed *seedmanagementv1alpha1.ManagedSeed, shoot *gardencorev1beta1.Shoot, checkOwner bool) error {
+	// Get seed object
+	seed := &gardencorev1beta1.Seed{}
+	if err := a.gardenClient.Client().Get(ctx, kutil.Key(managedSeed.Name), seed); err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil
+		}
+		return err
+	}
+
+	// Delete seed object
+	if err := a.gardenClient.Client().Delete(ctx, seed); client.IgnoreNotFound(err) != nil {
+		return err
+	}
+
+	// Check if there are any remaining objects associated with this seed
+	if err := checkSeedAssociations(ctx, a.gardenClient.Client(), seed.Name); err != nil {
```

Comment:
This check should be part of a admission plugin, which will prevent the deletion of the `ManagedSeed` in the first place, if there are still resources scheduled to that seed.
Basically, [this](https://github.com/gardener/gardener/blob/ec5794c5fddf09388f4e976f52327fd3d3d152b1/plugin/pkg/shoot/validator/admission.go#L912) code but only for `ManagedSeeds`.

Reply:
OK, makes sense, but then do you have an idea why this was present in the old implementation? https://github.com/gardener/gardener/blob/5a58c19bcd708bb1566e494cb7f08d479bc3cd08/pkg/gardenlet/controller/shoot/seed_registration_control.go#L219

Reply:
Hmm, it seems I misunderstood this admission check. It just denies removal of the `use-as-seed` annotation, but doesn't actually deny deletion of a Shoot with the annotation.
Then the controller goes ahead and just marks the Seed object for deletion and waits for all Shoots/BBs/BEs to be gone.
Probably, we can also keep this for now.

Though, I would actually like to see a cleaner way of implementing the deletion process somewhen: 
First mark the Seed as unschedulable, then migrate all remaining Shoots away (once that's implemented, otherwise just wait) and only delete the Seed object afterwards. Basically this would mean, the `ManagedSeed` controller would drain the Seed and then delete it.

Reply:
OK, I will keep the check for now.